lib/view_component/cacheable.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module ViewComponent::Cacheable
+  extend ActiveSupport::Concern
+
+  included do
+    class_attribute :__vc_cache_dependencies, default: []
+
+    # For caching, such as #cache_if
+    #
+    # @private
+    def view_cache_dependencies
+      return if __vc_cache_dependencies.blank? || __vc_cache_dependencies.none?
+
+      __vc_cache_dependencies.filter_map { |dep| send(dep) }
+    end
+
+    # For handeling the output_preamble and output_postamble
+    #
+    # @private
+    def __vc_render_template(rendered_template)
+      # Avoid allocating new string when output_preamble and output_postamble are blank
+      if output_preamble.blank? && output_postamble.blank?
+        rendered_template
+      else
+        safe_output_preamble + rendered_template + safe_output_postamble
+      end
+    end
+
+    # For determing if a template is rendered with cache or not
+    #
+    # @private
+    def __vc_render_cacheable(rendered_template)
+      if view_cache_dependencies.present?
+        Rails.cache.fetch(view_cache_dependencies) do
+          __vc_render_template(rendered_template)
+        end
+      else
+        __vc_render_template(rendered_template)
+      end
+    end
+  end
+
+  class_methods do
+    # For caching the component
+    def cache_on(*args)
+      __vc_cache_dependencies.push(*args)
+    end
+
+    def inherited(child)
```

Comment:
We took this for a spin in a few cases of our cached partials and view components that we are familiar with and largely it works for the base cases. This is working to correctly bust a cached VC when it changes or things in the render path 'above' (partials or VCs) it change.

What we aren't seeing is when a VC renders another VC or partial as a child that downstream changes of the cached VC pick up changes and handle it. I'm going to fork your branch and offer some test cases based on an approach we use in the [vc fragment caching gem](https://github.com/patrickarnett/view_component-fragment_caching) (what we're currently solving this with)

Effectively we want some way for a child partial or VC change (in either the `.rb` or template) to bust the cached parent.

Reply:
@JWShuff like touch: true in rails ?

Reply:
also big thanks for testing it, I can look at adding the touch true thing later this week